reward/addressBookManager.go

```diff
@@ -0,0 +1,186 @@
+	}
+
+	// Parse and construct node information
+	nodeIds := []common.Address{}
+	stakingAddrs := []common.Address{}
+	rewardAddrs := []common.Address{}
+	pocAddr := common.Address{}
+	kirAddr := common.Address{}
+	for i, addrType := range *allTypeList {
+		switch addrType {
+		case addressTypeNodeID:
+			nodeIds = append(nodeIds, (*allAddressList)[i])
+		case addressTypeStakingAddr:
+			stakingAddrs = append(stakingAddrs, (*allAddressList)[i])
+		case addressTypeRewardAddr:
+			rewardAddrs = append(rewardAddrs, (*allAddressList)[i])
+		case addressTypePoCAddr:
+			pocAddr = (*allAddressList)[i]
+		case addressTypeKIRAddr:
+			kirAddr = (*allAddressList)[i]
+		default:
+			return nil, nil, nil, common.Address{}, common.Address{}, errors.New(fmt.Sprintf("invalid type from AddressBook: %d", addrType))
+		}
+	}
+
+	// validate parsed node information
+	if len(nodeIds) != len(stakingAddrs) ||
+		len(nodeIds) != len(rewardAddrs) ||
+		isEmptyAddress(pocAddr) ||
+		isEmptyAddress(kirAddr) {
+		return nil, nil, nil, common.Address{}, common.Address{}, errAddressBookIncomplete
+	}
+
+	return nodeIds, stakingAddrs, rewardAddrs, pocAddr, kirAddr, nil
+}
+
+// getStakingInfoFromAddressBook returns staking info when calling AddressBook
+// succeeded. It returns an error otherwise.
+func (abm *addressBookManager) getStakingInfoFromAddressBook(blockNum uint64) (*StakingInfo, error) {
+	var nodeIds []common.Address
+	var stakingAddrs []common.Address
+	var rewardAddrs []common.Address
+	var KIRAddr = common.Address{}
+	var PoCAddr = common.Address{}
+	var err error
+
+	if !params.IsStakingUpdateInterval(blockNum) {
+		return nil, errors.New(fmt.Sprintf("not staking block number. blockNum: %d", blockNum))
+	}
+
+	// Prepare a message
+	msg, err := abm.makeMsgToAddressBook()
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("failed to make message for AddressBook. root err: %s", err))
+	}
+
+	intervalBlock := abm.bc.GetBlockByNumber(blockNum)
+	if intervalBlock == nil {
+		return nil, errors.New("stateDB is not ready for staking info")
+	}
+	statedb, err := abm.bc.StateAt(intervalBlock.Root())
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("failed to make a state for interval block. blockNum: %d, root err: %s", blockNum, err))
+	}
+
+	// Create a new context to be used in the EVM environment
+	context := blockchain.NewEVMContext(msg, intervalBlock.Header(), abm.bc, nil)
+	evm := vm.NewEVM(context, statedb, abm.chainConfig, &vm.Config{})
+
+	res, gas, kerr := blockchain.ApplyMessage(evm, msg)
+	logger.Trace("Call AddressBook contract", "used gas", gas, "kerr", kerr)
+	err = kerr.ErrTxInvalid
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("failed to call AddressBook contract. root err: %s", err))
+	}
+
+	nodeIds, stakingAddrs, rewardAddrs, PoCAddr, KIRAddr, err = abm.getAllAddressFromAddressBook(res)
```

Comment:
How about this?

nodeIds, stakingAddrs, rewardAddrs, PoCAddr, KIRAddr, err `:=` abm.getAllAddressFromAddressBook(res)

Now, we can remove `var nodeIds []common.Address ...` at the top of this method.

Reply:
reflect